executor/aggfuncs/func_variance.go

```diff
@@ -0,0 +1,427 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggfuncs
+
+import (
+	"math"
+
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/types"
+	"github.com/pingcap/tidb/util/chunk"
+	"github.com/pingcap/tidb/util/set"
+	"github.com/pkg/errors"
+)
+
+type varianceType int
+
+const (
+	varianceNone varianceType = iota
+	varianceVarPop
+	varianceVarSamp
+	varianceStdPop
+	varianceStdSamp
+)
+
+// "baseVarianceDecimal" is wrapped by:
+// - "varianceOriginal4Decimal"
+// - "variancePartial4Decimal"
+type baseVarianceDecimal struct {
+	baseAggFunc
+	varianceType
+}
+
+type partialResult4VarianceDecimal struct {
+	count    int64
+	sum      types.MyDecimal
+	variance types.MyDecimal
+}
+
+func (p *partialResult4VarianceDecimal) merge(count int64, sum, variance types.MyDecimal) error {
+	if p.count == 0 {
+		p.count, p.sum, p.variance = count, sum, variance
+	} else {
+		variance, sum, err := types.CalculateMergeDecimal(p.count, count, &p.sum, &sum, &p.variance, &variance)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		p.variance = *variance
+		p.sum = *sum
+		p.count += count
+	}
+	return nil
+}
+
+func (e *baseVarianceDecimal) AllocPartialResult() PartialResult {
+	return PartialResult(&partialResult4VarianceDecimal{})
+}
+
+func (e *baseVarianceDecimal) ResetPartialResult(pr PartialResult) {
+	p := (*partialResult4VarianceDecimal)(pr)
+	p.sum, p.variance = *types.NewDecFromInt(0), *types.NewDecFromInt(0)
+	p.count = int64(0)
+}
+
+func (e *baseVarianceDecimal) AppendFinalResult2Chunk(sctx sessionctx.Context, pr PartialResult, chk *chunk.Chunk) error {
+	p := (*partialResult4VarianceDecimal)(pr)
+	if p.count == 0 {
```

Comment:
It's better to move this check into the `switch...case` for `varianceStdPop` and `varianceVarPop`